skbio/tree/_nj.py

```diff
@@ -86,209 +112,232 @@ def nj(dm, disallow_negative_branch_length=True, result_constructor=None):
+
+        # Update cluster indices. Specifically, position i will have the new cluster
+        # index. Meanwhile, position j will be replaced with the last cluster.
+        idxs_[i] = 2 * N - n
+        idxs_[j] = idxs_[n_1]
+
+        n -= 1
+
+    # Perform final calculation on the three remaining taxa. They will become children
+    # of the root node, and the entire tree is unrooted.
+    L_0 = (dm[0, 1] + dm[0, 2] - dm[1, 2]) / 2
+    lm[N - 3] = idxs[1], idxs[2], dm[0, 1] - L_0, dm[0, 2] - L_0
+    lm[N - 2] = idxs[0], 2 * N - 3, L_0, 0
+
+    return lm
+
+
+def _tree_from_linkmat(lm, taxa, rooted=True, clip_to_zero=True):
+    r"""Convert a linkage matrix into a tree structure.
 
     Parameters
     ----------
-    dm : skbio.DistanceMatrix
-        The input distance matrix.
-    i, j : str
-        Identifiers of entries in the distance matrix to be collapsed (i.e.,
-        the descendants of the new node, which is internally represented as
-        `u`).
-    disallow_negative_branch_length : bool
-        Neighbor joining can result in negative branch lengths, which don't
-        make sense in an evolutionary context. If `True`, negative branch
-        lengths will be returned as zero, a common strategy for handling this
-        issue that was proposed by the original developers of the algorithm.
+    lm : (N, 4) array_like
+        Linkage matrix.
+    taxa : list of str
+        Taxon names.
+    rooted : bool, optional
+        If True (default), will generate a rooted binary tree, with two children
+        attached to the root node. If False, will generate an unrooted tree with
+        three children attached to the root node.
+    clip_to_zero : bool, optional
+        Convert negative branch lengths to zero.
 
-    """
-    n = dm.shape[0]
-    i_to_j = dm[i, j]
-    i_to_u = (0.5 * i_to_j) + ((dm[i].sum() - dm[j].sum()) / (2 * (n - 2)))
+    Returns
+    -------
+    TreeNode
+        Converted phylogenetic tree.
 
-    if disallow_negative_branch_length and i_to_u < 0:
-        i_to_u = 0
+    See Also
+    --------
+    TreeNode.from_linkage_matrix
 
-    j_to_u = i_to_j - i_to_u
+    Notes
+    -----
+    The linkage matrix data structure resembles that used in SciPy's hierarchical
+    clustering, but also stores the branch lengths of both children, which could be
+    unequal in a phylogenetic tree.
+
+    In SciPy's linkage matrix, the four elements per row are (see:
+    https://stackoverflow.com/questions/9838861/):
+
+        taxon1, taxon2, length1+2, # original taxa
 
-    if disallow_negative_branch_length and j_to_u < 0:
-        j_to_u = 0
+    In the current data structure, they are:
 
-    return i_to_u, j_to_u
+        taxon1, taxon2, length1, length2
+
+    """
+    # allocate node list
+    nodes = [TreeNode(name=x) for x in taxa] + [TreeNode() for _ in range(len(lm))]
+
+    # build tree incrementally
+    idx = len(taxa)
+    for c1, c2, l1, l2 in lm if rooted else lm[:-2]:
```

Comment:
`l1`, `l2` are hard to read in my opinion. They look very close to the numbers `11` and `12`. I'd suggest naming the variables something more descriptive or that avoids the use of a lot of lowercase `l`'s. A simple change to capital `L` would also suffice.